File: backend/app/services/onboarding_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from typing import Optional, List
import json

from app.models.user import User, UserProfile, UserConsent
from app.schemas.onboarding import (
    OnboardingGoalsRequest, 
    OnboardingProfileRequest, 
    OnboardingConsentRequest,
    OnboardingStatusResponse,
    OnboardingCompleteResponse
)

class OnboardingService:
    
    def get_onboarding_status(self, db: Session, user_id: str) -> OnboardingStatusResponse:
        """사용자의 온보딩 상태 조회"""
        user_profile = db.query(UserProfile).filter(UserProfile.user_id == user_id).first()
        
        if not user_profile:
            return OnboardingStatusResponse(
                onboarding_completed=False,
                profile_completion_rate=0,
                current_step="goals"
            )
        
        # 단계별 완성도 계산
        completion_rate = 0
        current_step = "goals"
        
        # 목표 설정 완료 여부 (goals 필드 체크)
        if user_profile.goals:
            completion_rate += 25
            current_step = "profile"
        
        # 기본 정보 완료 여부 (age, region 필수)
        if user_profile.age and user_profile.region:
            completion_rate += 25
            current_step = "consent"
        
        # 동의 정보 완료 여부
        user_consent = db.query(UserConsent).filter(UserConsent.user_id == user_id).first()
        if user_consent and user_consent.privacy_policy and user_consent.terms_of_service:
            completion_rate += 25
            current_step = "complete"
        
        # 온보딩 완료 여부
        onboarding_completed = user_profile.onboarding_completed or completion_rate >= 75
        if onboarding_completed:
            completion_rate = 100
            current_step = "completed"
            
        return OnboardingStatusResponse(
            onboarding_completed=onboarding_completed,
            profile_completion_rate=completion_rate,
            current_step=current_step
        )
    
    def save_goals(self, db: Session, user_id: str, goals_request: OnboardingGoalsRequest) -> dict:
        """온보딩 목표 저장"""
        try:
            user_profile = db.query(UserProfile).filter(UserProfile.user_id == user_id).first()
            
            if not user_profile:
                # UserProfile이 없으면 생성
                user_profile = UserProfile(
                    user_id=user_id,
                    goals=json.dumps(goals_request.goals),
                    profile_completion_rate=25
                )
                db.add(user_profile)
            else:
                # 기존 프로필 업데이트
                user_profile.goals = json.dumps(goals_request.goals)
                # 완성도 재계산
                if user_profile.age and user_profile.region:
                    user_profile.profile_completion_rate = 50
                else:
                    user_profile.profile_completion_rate = 25
            
            db.commit()
            db.refresh(user_profile)
            
            return {"success": True, "message": "목표가 성공적으로 저장되었습니다."}
            
        except Exception as e:
            db.rollback()
            return {"success": False, "error": str(e)}
    
    def save_profile(self, db: Session, user_id: str, profile_request: OnboardingProfileRequest) -> dict:
        """온보딩 프로필 정보 저장"""
        try:
            user_profile = db.query(UserProfile).filter(UserProfile.user_id == user_id).first()
            
            if not user_profile:
                # UserProfile이 없으면 생성
                user_profile = UserProfile(
                    user_id=user_id,
                    age=profile_request.age,
                    region=profile_request.region,
                    job_category=profile_request.job_category,
                    income_range=profile_request.income_range,
                    profile_completion_rate=25
                )
                db.add(user_profile)
            else:
                # 기존 프로필 업데이트
                user_profile.age = profile_request.age
                user_profile.region = profile_request.region
                user_profile.job_category = profile_request.job_category
                user_profile.income_range = profile_request.income_range
                
                # 완성도 재계산
                completion_rate = 25  # 기본 정보
                if user_profile.goals:
                    completion_rate += 25  # 목표
                user_profile.profile_completion_rate = completion_rate
            
            db.commit()
            db.refresh(user_profile)
            
            return {"success": True, "message": "프로필 정보가 성공적으로 저장되었습니다."}
            
        except Exception as e:
            db.rollback()
            return {"success": False, "error": str(e)}
    
    def save_consent(self, db: Session, user_id: str, consent_request: OnboardingConsentRequest) -> dict:
        """온보딩 동의 정보 저장"""
        try:
            user_consent = db.query(UserConsent).filter(UserConsent.user_id == user_id).first()
            
            if not user_consent:
                # UserConsent가 없으면 생성
                user_consent = UserConsent(
                    user_id=user_id,
                    push_notification=consent_request.push_notification,
                    marketing_notification=consent_request.marketing_notification,
                    reward_program=consent_request.reward_program,
                    privacy_policy=consent_request.privacy_policy,
                    terms_of_service=consent_request.terms_of_service
                )
                db.add(user_consent)
            else:
                # 기존 동의 정보 업데이트
                user_consent.push_notification = consent_request.push_notification
                user_consent.marketing_notification = consent_request.marketing_notification
                user_consent.reward_program = consent_request.reward_program
                user_consent.privacy_policy = consent_request.privacy_policy
                user_consent.terms_of_service = consent_request.terms_of_service
            
            # 프로필 완성도 업데이트
            user_profile = db.query(UserProfile).filter(UserProfile.user_id == user_id).first()
            if user_profile:
                completion_rate = 75  # 기본 정보 + 목표 + 동의
                user_profile.profile_completion_rate = completion_rate
            
            db.commit()
            
            return {"success": True, "message": "동의 정보가 성공적으로 저장되었습니다."}
            
        except Exception as e:
            db.rollback()
            return {"success": False, "error": str(e)}
    
    def complete_onboarding(self, db: Session, user_id: str) -> OnboardingCompleteResponse:
        """온보딩 완료 처리"""
        try:
            user_profile = db.query(UserProfile).filter(UserProfile.user_id == user_id).first()
            
            if not user_profile:
                return OnboardingCompleteResponse(
                    success=False,
                    message="사용자 프로필을 찾을 수 없습니다.",
                    user_id=user_id
                )
            
            # 온보딩 완료 표시
            user_profile.onboarding_completed = True
            user_profile.profile_completion_rate = 100
            
            db.commit()
            db.refresh(user_profile)
            
            return OnboardingCompleteResponse(
                success=True,
                message="온보딩이 성공적으로 완료되었습니다!",
                user_id=user_id
            )
            
        except Exception as e:
            db.rollback()
            return OnboardingCompleteResponse(
                success=False,
                message=f"온보딩 완료 처리 중 오류가 발생했습니다: {str(e)}",
                user_id=user_id
            )

onboarding_service = OnboardingService()

File: FinKuRN/src/services/onboardingService.ts
/**
 * 온보딩 서비스 (Onboarding Service)
 *
 * 온보딩 프로세스의 모든 단계를 관리하는 서비스입니다.
 * 사용자의 목표, 기본 정보, 동의 사항을 백엔드에 저장하고
 * 온보딩 상태를 추적합니다.
 *
 * @module services/onboardingService  
 * @category Services
 * @since 1.0.0
 */

import type { 
  UserGoal, 
  JobCategory, 
  IncomeRange,
  GoalSelectionData,
  BasicInfoData,
  ConsentData 
} from '../types/onboarding';

const API_BASE_URL = 'http://localhost:8001/api/v1';

/**
 * 온보딩 상태 응답 인터페이스
 */
export interface OnboardingStatusResponse {
  onboarding_completed: boolean;
  profile_completion_rate: number;
  current_step: string;
}

/**
 * 기본 API 응답 인터페이스
 */
export interface ApiResponse {
  success: boolean;
  message?: string;
  error?: string;
}

/**
 * 온보딩 완료 응답 인터페이스
 */
export interface OnboardingCompleteResponse extends ApiResponse {
  user_id: string;
}

/**
 * 인증 토큰 가져오기 (AsyncStorage에서)
 */
const getAuthToken = async (): Promise<string | null> => {
  try {
    // React Native 환경에서 AsyncStorage 사용
    if (typeof window === 'undefined') {
      const AsyncStorage = require('@react-native-async-storage/async-storage').default;
      return await AsyncStorage.getItem('authToken');
    }
    
    // 웹 환경에서는 localStorage 사용
    return localStorage.getItem('authToken');
  } catch (error) {
    console.error('토큰 가져오기 실패:', error);
    return null;
  }
};

/**
 * API 헤더 생성
 */
const getApiHeaders = async () => {
  const token = await getAuthToken();
  return {
    'Content-Type': 'application/json',
    ...(token && { 'Authorization': `Bearer ${token}` })
  };
};

/**
 * 온보딩 상태 조회
 */
export const getOnboardingStatus = async (): Promise<OnboardingStatusResponse | null> => {
  try {
    const headers = await getApiHeaders();
    
    const response = await fetch(`${API_BASE_URL}/onboarding/status`, {
      method: 'GET',
      headers,
    });

    if (!response.ok) {
      throw new Error(`API 오류: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('온보딩 상태 조회 실패:', error);
    return null;
  }
};

/**
 * 목표 선택 저장
 */
export const saveGoals = async (goals: UserGoal[]): Promise<ApiResponse> => {
  try {
    const headers = await getApiHeaders();

    const response = await fetch(`${API_BASE_URL}/onboarding/goals`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ goals }),
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.error || `API 오류: ${response.status}`);
    }

    // 백엔드가 {message: "..."} 형식으로 반환하므로 success: true 추가
    return {
      success: true,
      message: data.message || '목표가 저장되었습니다.'
    };
  } catch (error) {
    console.error('목표 저장 실패:', error);
    return {
      success: false,
      error: error instanceof Error ? error.message : '목표 저장에 실패했습니다.'
    };
  }
};

/**
 * 기본 정보 저장
 */
export const saveProfile = async (profile: BasicInfoData): Promise<ApiResponse> => {
  try {
    const headers = await getApiHeaders();

    const requestData = {
      age: profile.age,
      region: profile.region,
      job_category: profile.jobCategory,
      income_range: profile.incomeRange,
    };

    const response = await fetch(`${API_BASE_URL}/onboarding/profile`, {
      method: 'POST',
      headers,
      body: JSON.stringify(requestData),
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.error || `API 오류: ${response.status}`);
    }

    // 백엔드가 {message: "..."} 형식으로 반환하므로 success: true 추가
    return {
      success: true,
      message: data.message || '기본 정보가 저장되었습니다.'
    };
  } catch (error) {
    console.error('기본 정보 저장 실패:', error);
    return {
      success: false,
      error: error instanceof Error ? error.message : '기본 정보 저장에 실패했습니다.'
    };
  }
};

/**
 * 동의 사항 저장
 */
export const saveConsent = async (consent: ConsentData): Promise<ApiResponse> => {
  try {
    const headers = await getApiHeaders();
    
    const requestData = {
      push_notification: consent.pushNotification,
      marketing_notification: consent.marketingNotification, 
      reward_program: consent.rewardProgram,
    };
    
    const response = await fetch(`${API_BASE_URL}/onboarding/consent`, {
      method: 'POST',
      headers,
      body: JSON.stringify(requestData),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.error || `API 오류: ${response.status}`);
    }

    return data;
  } catch (error) {
    console.error('동의 사항 저장 실패:', error);
    return {
      success: false,
      error: error instanceof Error ? error.message : '동의 사항 저장에 실패했습니다.'
    };
  }
};

/**
 * 온보딩 완료
 */
export const completeOnboarding = async (): Promise<OnboardingCompleteResponse | null> => {
  try {
    const headers = await getApiHeaders();
    
    const response = await fetch(`${API_BASE_URL}/onboarding/complete`, {
      method: 'POST',
      headers,
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.error || `API 오류: ${response.status}`);
    }

    return data;
  } catch (error) {
    console.error('온보딩 완료 처리 실패:', error);
    return null;
  }
};

/**
 * 온보딩 서비스 객체 Export
 */
export const onboardingService = {
  getOnboardingStatus,
  saveGoals,
  saveProfile,
  saveConsent,
  completeOnboarding,
};